src/clr-angular/data/datagrid/providers/state.provider.ts

```diff
@@ -16,111 +16,162 @@ import { FiltersProvider } from './filters';
+            for (const filter of activeFilters) {
+                if (filter.filterState && filter.filterState.type === 'BuiltinStringFilter') {
+                    const stringFilterState = <StringFilterStateInterface>filter.filterState;
+                    state.filters.push({
+                        property: stringFilterState.property,
+                        value: stringFilterState.value,
+                    });
+                } else if (filter.filterState && filter.filterState.type) {
+                    state.filters.push(filter);
+                }
+            }
         }
-        state.filters.push(filter);
-      }
+        return state;
     }
-    return state;
-  }
-
-  set state(state: ClrDatagridStateInterface<T>) {
-    this.debouncer.changeStart();
-    if (state.page) {
-      this.page.size = state.page.size;
-      this.page.current = Math.ceil(state.page.from / state.page.size) + 1;
+
+    set state(state: ClrDatagridStateInterface<T>) {
+        if (this.sameAsPreviousState(state)) {
+            return;
+        } else {
+            this._prevState = state;
+        }
+        this.debouncer.changeStart();
+        if (state.page) {
+            this.page.size = state.page.size;
+            this.page.current = Math.ceil(state.page.from / state.page.size) + 1;
+        }
+        if (state.sort) {
+            if (typeof state.sort.by === 'string') {
+                if (!(this.sort.comparator instanceof DatagridPropertyComparator)) {
+                    this.sort.comparator = new DatagridPropertyComparator(state.sort.by);
+                }
+                if (this.sort.reverse !== state.sort.reverse) {
+                    this.sort.toggle(this.sort.comparator);
+                }
+            } else {
+                this.sort.comparator = state.sort.by;
+                this.sort.reverse = state.sort.reverse;
+            }
+        }
+        if (state.filters) {
+            const activeFilters = this.filters.getActiveFilters();
+            for (const filter of state.filters) {
+                let filterObject: SerializableFilter<T>;
+                if (filter.hasOwnProperty('property') && filter.hasOwnProperty('value')) {
+                    const defaultFilterRepresentation = filter as { property: string; value: string };
+                    const propertyStringFilter = new DatagridPropertyStringFilter(defaultFilterRepresentation.property);
+                    const stringFilter = new DatagridStringFilterImpl(propertyStringFilter);
+                    stringFilter.value = defaultFilterRepresentation.value;
+                    filterObject = stringFilter;
+                } else {
+                    filterObject = filter as SerializableFilter<T>;
+                }
+                const existing = activeFilters.findIndex(value => value.compatibleToState(filterObject.filterState));
+                if (existing !== -1) {
+                    activeFilters.splice(existing, 1);
```

Comment:
At this point we only found a compatible filter, but now we must set the state of that filter with `existing.filterState = filter`.